### PR TITLE
[receiver/sqlquery] Add support of optional Start and Stop Timestamp

### PR DESCRIPTION
**Pull request by Pranav-SA**
In this PR, support for Start and Stop Timestamp to be picked up from the metrics stored in the rows is added. This will allow SQL queries to be more expressive and allows users to define the period of aggregation which is not only dependent on the default system time set upon processing.

**Link to tracking Issue:** 
#18925 
#14146

### CODE REVIEW

File: receiver/sqlqueryreceiver/README.md [hunk 1]
```diff
@@ -37,8 +37,9 @@ Each _metric_ in the configuration will produce one OTel metric per row returned
 * `aggregation` (optional): only applicable for `data_type=sum`; can be `cumulative` or `delta`; defaults to `cumulative`.
 * `description` (optional): the description applied to the metric.
 * `unit` (optional): the units applied to the metric.
-* `static_attributes` (optional): static attributes applied to the metrics
-
+* `static_attributes` (optional): static attributes applied to the metrics.
+* `start_ts_column` (optional): start timestamp associated with metrics.
+* `stop_ts_column` (optional): stop timestamp associated with metrics, used for setting metric `timestamp`; defaults to time of processing.
```
Comment (atoulme) [inline receiver/sqlqueryreceiver/README.md:42]: I don’t see where it defaults to time of processing.

Comment (Pranav-SA) [reply receiver/sqlqueryreceiver/README.md]: Thanks for pointing that out! I've edited the Readme accordingly.

File: receiver/sqlqueryreceiver/config.go [hunk 2]
```diff
@@ -103,6 +106,16 @@ func (c MetricCfg) Validate() error {
 	if c.DataType == MetricTypeGauge && c.Aggregation != "" {
 		errs = multierr.Append(errs, fmt.Errorf("aggregation=%s but data_type=%s does not support aggregation", c.Aggregation, c.DataType))
 	}
+	if c.StartTsColumn != "" {
+		if _, err := strconv.ParseUint(c.StartTsColumn, 10, 64); err != nil {
```
Comment (astencel-sumo) [inline receiver/sqlqueryreceiver/config.go:110]: Why are you trying to parse column names into integers? This doesn't look right. My understanding is that the `start_ts_column` and `stop_ts_column` can contain a string, which is the name of a column. Trying to parse this into a string doesn't make much sense.

File: receiver/sqlqueryreceiver/config.go [hunk 3]
```diff
@@ -81,6 +82,8 @@ type MetricCfg struct {
 	Unit             string            `mapstructure:"unit"`
 	Description      string            `mapstructure:"description"`
 	StaticAttributes map[string]string `mapstructure:"static_attributes"`
+	StartTsColumn    string            `mapstructure:"start_ts_column"`
+	StopTsColumn     string            `mapstructure:"stop_ts_column"`
```
Comment (astencel-sumo) [inline receiver/sqlqueryreceiver/config.go:86]: Are you sure that `stop_ts_column` is the right name for this setting? Isn't it more like a "current" timestamp?

Comment (Pranav-SA) [reply receiver/sqlqueryreceiver/config.go:86]: Depending upon the received rows stop may not necessarily be the current timestamp. I agree `stop_ts_column` is not the right name.  I am changing it to `end_ts_column`.

File: receiver/sqlqueryreceiver/config.go (hunk 2 above)
Comment (Pranav-SA) [reply receiver/sqlqueryreceiver/config.go]: Sincere apologies for the confusion. You're right this is not required. Omitting it.

File: receiver/sqlqueryreceiver/metrics.go [hunk 4]
```diff
@@ -29,6 +29,24 @@ func rowToMetric(row stringMap, cfg MetricCfg, dest pmetric.Metric, startTime pc
 	dest.SetUnit(cfg.Unit)
 	dataPointSlice := setMetricFields(cfg, dest)
 	dataPoint := dataPointSlice.AppendEmpty()
+	if cfg.StartTsColumn != "" {
+		if val, found := row[cfg.StartTsColumn]; found {
+			timestamp, err := strconv.ParseInt(val, 10, 64)
+			if err != nil {
+				return fmt.Errorf("failed to parse uint64 for %s, value was %s: %w", cfg.StartTsColumn, val, err)
```
Comment (jpkrohling) [inline receiver/sqlqueryreceiver/metrics.go:36]:
```suggestion
				return fmt.Errorf("failed to parse uint64 for %q, value was %q: %w", cfg.StartTsColumn, val, err)
```

File: receiver/sqlqueryreceiver/metrics.go [hunk 5]
```diff
@@ -29,6 +29,24 @@ func rowToMetric(row stringMap, cfg MetricCfg, dest pmetric.Metric, startTime pc
...
+			}
+			startTime = pcommon.Timestamp(timestamp)
+		}
+	}
+	if cfg.EndTsColumn != "" {
+		if val, found := row[cfg.EndTsColumn]; found {
+			timestamp, err := strconv.ParseInt(val, 10, 64)
+			if err != nil {
+				return fmt.Errorf("failed to parse uint64 for %s, value was %s: %w", cfg.EndTsColumn, val, err)
```
Comment (jpkrohling) [inline receiver/sqlqueryreceiver/metrics.go:45]:
```suggestion
				return fmt.Errorf("failed to parse uint64 for %q, value was %q: %w", cfg.EndTsColumn, val, err)
```

File: receiver/sqlqueryreceiver/metrics.go [hunk 6]
```diff
@@ -29,6 +29,24 @@ func rowToMetric(row stringMap, cfg MetricCfg, dest pmetric.Metric, startTime pc
 	dest.SetUnit(cfg.Unit)
 	dataPointSlice := setMetricFields(cfg, dest)
 	dataPoint := dataPointSlice.AppendEmpty()
+	if cfg.StartTsColumn != "" {
+		if val, found := row[cfg.StartTsColumn]; found {
```
Comment (jpkrohling) [inline receiver/sqlqueryreceiver/metrics.go:33]: Shouldn't this fail in case the column isn't found?

Comment (astencel-sumo) [reply receiver/sqlqueryreceiver/metrics.go:33]: Right, thanks for pointing this out Juraci. I agree this should fail when the column is not found, to make it consistent with the behavior of `attribute_columns` property, which also fails when the column is not found for an attribute.

**Review by andrzej-stencel — APPROVED** [review]

File: receiver/sqlqueryreceiver/metrics.go [hunk 7]
```diff
@@ -29,6 +29,28 @@ func rowToMetric(row stringMap, cfg MetricCfg, dest pmetric.Metric, startTime pc
...
+			timestamp, err := strconv.ParseInt(val, 10, 64)
```
Comment (pmcollins) [inline receiver/sqlqueryreceiver/metrics.go:34]: Can we replace these numbers with constants (here and on line 45)?

**Review by pmcollins — CHANGES_REQUESTED** [review]

File: receiver/sqlqueryreceiver/metrics.go (hunk 7 above)
Comment (Pranav-SA) [reply receiver/sqlqueryreceiver/metrics.go:34]: The following has been used to keep it consistent with other similar scenarios. Also, the usage is confined to only two places.

**Review by pmcollins — APPROVED** [review]

**Review by andrzej-stencel — APPROVED** [review]
Awesome! Thanks a lot for this contribution @Pranav-SA 🙏

**Comment by astencel-sumo** [comment]
The [build-and-test / build-package (deb)](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4811061199/jobs/8788272259?pr=19160) job has failed, but I don't think it's related to this PR.

@jpkrohling @dmitryax do we need anything more before merging this PR?

**Review by codeboten — COMMENTED** [review]
Thanks for the patience @Pranav-SA, looks like a couple of conflicts need resolving, once that's done it can be merged

**Review by jpkrohling — APPROVED** [review]

**Comment by devshashankb** [comment]
@Pranav-SA Can you please help with this issue. i am trying to use ts_column, but i don't see the metric having the timestamp from the db. May be i am missing something on the configuration 

https://github.com/open-telemetry/opentelemetry-collector-contrib/discussions/26081

**Comment by astencel-sumo** [comment]
@devshashankb I've answered in the discussion https://github.com/open-telemetry/opentelemetry-collector-contrib/discussions/26081#discussioncomment-7313801, please let me know if it works for you.

**Pull request merged**